include/caffe/layer.hpp

```diff
@@ -316,6 +317,21 @@ class Layer {
     param_propagate_down_[param_id] = value;
   }
 
+  bool IsForwardPassed() const {
```

Comment:
This is part of the annoyance of Caffe's lazy allocation.  There are allocations that happen late, specifically setup of cuRand, so we need to be able to delay final setup of algorithm choice until we know how much memory we really have.  (There is an upcoming PR to move to the findEx paths in cudnnv5 instead of get that make this even more challenging).  We think we need this more generally.  Would love a better solution, but we start going down a "plan" style path pretty quickly.